embassy-nrf/src/radio/ble.rs

```diff
@@ -0,0 +1,431 @@
+//! Radio driver implementation focused on Bluetooth Low-Energy transmission.
+//!
+//! The radio can calculate the CRC, perform data whitening,
+//! automatically send the right preamble.
+//! Most of the configuration is done automatically when you choose the mode and this driver.
+//!
+//! Some configuration can just be done when de device is disabled,
+//! and the configuration varies depending if is a transmitter or a receiver.
+//! Because of that we have a state machine to keep track of the state of the radio.
+//! The Radio is the disable radio which configure the common parameters between
+//! the bluetooth protocols, like the package format, the CRC and the whitening.
+//! The TxRadio radio enable and configured as a transmitter with the specific parameters.
+
+use core::future::poll_fn;
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::task::Poll;
+
+use embassy_hal_internal::drop::OnDrop;
+use embassy_hal_internal::{into_ref, PeripheralRef};
+pub use pac::radio::mode::MODE_A as Mode;
+use pac::radio::pcnf0::PLEN_A as PreambleLength;
+use pac::radio::state::STATE_A as RadioState;
+pub use pac::radio::txpower::TXPOWER_A as TxPower;
+
+use crate::interrupt::typelevel::Interrupt;
+use crate::radio::*;
+use crate::util::slice_in_ram_or;
+
+/// UART error.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+pub enum Error {
+    /// Buffer was too long.
+    BufferTooLong,
+    /// Buffer was to short.
+    BufferTooShort,
+    /// The buffer is not in data RAM. It is most likely in flash, and nRF's DMA cannot access flash.
+    BufferNotInRAM,
+}
+
+/// Radio driver.
+pub struct Radio<'d, T: Instance> {
+    _p: PeripheralRef<'d, T>,
+}
+
+impl<'d, T: Instance> Radio<'d, T> {
+    /// Create a new radio driver.
+    pub fn new(
+        radio: impl Peripheral<P = T> + 'd,
+        _irq: impl interrupt::typelevel::Binding<T::Interrupt, InterruptHandler<T>> + 'd,
+    ) -> Self {
+        into_ref!(radio);
+
+        let r = T::regs();
+
+        r.pcnf1.write(|w| unsafe {
+            // It is 0 bytes long in a standard BLE packet
+            w.statlen()
+                .bits(0)
+                // MaxLen configures the maximum packet payload plus add-on size in
+                // number of bytes that can be transmitted or received by the RADIO. This feature can be used to ensure
+                // that the RADIO does not overwrite, or read beyond, the RAM assigned to the packet payload. This means
+                // that if the packet payload length defined by PCNF1.STATLEN and the LENGTH field in the packet specifies a
+                // packet larger than MAXLEN, the payload will be truncated at MAXLEN
+                //
+                // To simplify the implementation, I'm setting the max length to the maximum value
+                // and I'm using only the length field to truncate the payload
```

Comment:
This feels nitpicky, but try to not comment things as 'first person'. Also leave a TODO marker for things you think can be improved later, it makes it easier to find for newcomers to the code wanting to contribute.

Reply:
That makes complete sense, thank you very much for the feedback!